grabbit/src/main/groovy/com/twcable/grabbit/jcr/ProtoNodeDecorator.groovy

```diff
@@ -0,0 +1,90 @@
+package com.twcable.grabbit.jcr
+
+import com.twcable.grabbit.proto.NodeProtos.Node as ProtoNode
+import com.twcable.grabbit.proto.NodeProtos.Value as ProtoValue
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import org.apache.jackrabbit.commons.JcrUtils
+
+import javax.annotation.Nonnull
+import javax.jcr.Node as JCRNode
+import javax.jcr.Session
+
+import static org.apache.jackrabbit.JcrConstants.JCR_MIXINTYPES
+import static org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE
+
+@CompileStatic
+@Slf4j
+class ProtoNodeDecorator {
+
+    @Delegate
+    ProtoNode innerProtoNode
+
+    Collection<ProtoPropertyDecorator> protoProperties
+
+
+    ProtoNodeDecorator(@Nonnull ProtoNode node) {
+        if(!node) throw new IllegalArgumentException("node must not be null!")
+        this.innerProtoNode = node
+        this.protoProperties = node.properties.propertyList.collect { new ProtoPropertyDecorator(it) }
+    }
+
+
+    String getPrimaryType() {
+        protoProperties.find { it.isPrimaryType() }.value.stringValue
+    }
+
+
+    ProtoPropertyDecorator getMixinProperty() {
+        protoProperties.find { it.isMixinType() }
+    }
+
+
+    Collection<ProtoPropertyDecorator> getWritableProperties() {
+        protoProperties.findAll { !(it.name in [JCR_PRIMARYTYPE, JCR_MIXINTYPES]) }
+    }
+
+
+    JCRNodeDecorator writeToJcr(@Nonnull Session session) {
+        final jcrNode = getOrCreateNode(session)
+        //Write mixin types first to avoid InvalidConstraintExceptions
+        final mixinProperty = getMixinProperty()
+        if(mixinProperty) {
+            addMixins(mixinProperty, jcrNode)
+        }
+        //Then add other properties
+        writableProperties.each { it.writeToNode(jcrNode) }
+
+        return new JCRNodeDecorator(jcrNode)
+    }
+
+
+    /**
+     * This method is rather succinct, but helps isolate this JcrUtils static method call
+     * so that we can get better test coverage.
+     * @param session to create or get the node path for
+     * @return the newly created, or found node
+     */
+    JCRNode getOrCreateNode(Session session) {
+        JcrUtils.getOrCreateByPath(innerProtoNode.name, primaryType, session)
+    }
+
+
+    /**
+     * If a property can be added as a mixin, adds it to the given node
+     * @param property
+     * @param node
+     */
+    private static void addMixins(ProtoPropertyDecorator property, JCRNode node) {
+        property.values.valueList.each { ProtoValue value ->
```

Comment:
"10.4.2.4 No Null Values
Every property must have a value. The range of property states does not include having a “null value”, or “no value”. Setting a property to “null” is equivalent to removing that property (see §10.9 Removing Nodes and Properties)."